swap/src/bin/asb.rs

```diff
@@ -73,6 +78,21 @@ async fn main() -> Result<()> {
 
     let env_config = env::Testnet::get_config();
 
+    let uac = tor::UnauthenticatedConnection::default()
+        .with_control_port(config.tor.control_port)
+        .with_socks5_port(config.tor.socks5_port);
+    let (tor_socks5_port, _ac) = match uac.assert_tor_running().await {
+        Ok(_) => {
+            tracing::info!("Tor found. Setting up hidden service. ");
+            let ac = register_tor_services(config.network.clone().listen, uac).await?;
```

Comment:
> At this point in the program's lifecycle, you can't actually know whether or not you will be listening on these addresses.

I don't understand why not. You configure your service to listen on these addresses, why would it change?

> This snippet of code should either run:
> Here in the eventloop

The `eventloop` could work as well. However, ` SwarmEvent::NewListenAddr(addr)` will be fired for each local listening address. We should only care about `localhost`/`127.0.0.1`. 

> Like we discussed the other day, the latter would be the cleanest because it makes the whole Tor transport module self-contained.

I played with this thought but disagree with it. I think it would be weird if a network behavior has the side effect of interacting with Tor and adding hidden service. 


Personally I think the cleanest is to add hidden services from the outside as we have more information about the listening addresses we care about. 
Not everyone might be comfortable with handing over the control port to the application and with this approach, a user can configure the hidden service from the outside of the application. 
I used this approach successfully while testing and only added the feature of adding hidden Tor services later on.

Reply:
> I don't understand why not. You configure your service to listen on these addresses, why would it change?

It is not about whether or not it changes. It is about whether or not the _intent_ to listen on these addresses actually succeeds. `Swarm::listen_on` is asynchronous. Calling the function by itself doesn't do anything other than modifying some state. That state is picked up on the next `poll` of the Swarm and the transport is instructed to _try_ and listen on the given multi-address.

You are not actually listening on the interface until the underlying transport emits `ListenerEvent::NewAddress`.

> However, SwarmEvent::NewListenAddr(addr) will be fired for each local listening address. We should only care about localhost/127.0.0.1.

There is `is_loopback()`: https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#method.is_loopback

> I played with this thought but disagree with it. I think it would be weird if a network behavior has the side effect of interacting with Tor and adding hidden service.

I agree that a `NetworkBehaviour` would be weird. I am talking about the `Transport` though. A `Transport`'s responsibility is to _dial_ and _listen_ using a certain transport protocol. If you have a Tor-`Transport`, it seems perfectly reasonable that the implementation somehow achieves that it is reachable via Tor. One way of doing this is setting up an ephemeral service. There are other ways as well so we will probably need some configuration at some point.

On the point of side-effects: A `NetworkBehaviour` as well as a `Transport` need to be **explicitly** included by the user in the configuration. It doesn't appear out of nowhere. As part of choosing to include such a module, I would hope that they read the documentation and are aware of what provided functionality does.

> Personally I think the cleanest is to add hidden services from the outside as we have more information about the listening addresses we care about.

There seems to be a misunderstanding here. Like I explained above, you actually have _less_ information about the listen addresses on the outside. Already in-use ports are a problem for example. Another one is wildcard addresses like `0.0.0.0` that only get resolved to concrete interfaces by the kernel once you start listening. A third one might be unsupported multi-addresses like trying to listen on `/ip4/0.0.0.0/udp/3000` although you don't have a Udp transport configured.

To summarize, the `Transport` implementation has absolutely the most knowledge about what is going on in terms of which interfaces are being listened on etc. For example, the `Swarm` can only emit a `SwarmEvent::NewListenAddr` because the `Transport` emitted a `ListenerEvent::NewAddr` event.

> Not everyone might be comfortable with handing over the control port to the application and with this approach, a user can configure the hidden service from the outside of the application.

Happy to make it configurable then. From our discussion, I see at least three possible options:

- Don't listen on Tor at all, i.e. just pass through to the underlying TCP transport
- "Listen" on an existing hidden service. This one is a bit weird because effectively, the only thing you can do is emit a `ListenerEvent::NewAddr` and inform the rest of the application that we are listening on a particular onion address. I guess we could do some validation with the Tor node that the user set up Tor correctly to route to the port that we are actually listening on.
- Automatically set up one or more hidden service(s) for the addresses we are listening on. Can probably make this configurable to use random keys or existing ones.

---

In summary, having this code inside the `Transport` achieves multiple things:

- It is more correct. Unless you react to events of the `Swarm` about successful / failed listening, you don't know whether or not you are actually listening on an interface.
- It makes the functionality reusable because adding the transport is basically plug-and-play.
- Conceptually, a `Transport` is about listening and dialling. Having an implementation do only one of the two is a code-smell. The code-smell occurs again in the code inside `main` that sets up the hidden service that just sits there and doesn't know where it wants to belong.
- It makes the functionality independently testable. We can make a swarm that automatically sets up a hidden service, emits the correct listen address and we have another swarm that connects to it. That is how (to my knowledge) all `Transport`s are tested in rust-libp2p.

Hope this clears things up!

Reply:
So I thought about this in more detail and had some insights.

I came around to the view that a `Transport` shouldn't just magically do things. It should only do something once the user gave a clear instruction via `swarm.listen_on`. For example, a `TcpTransport` doesn't do anything unless the user says:

```
swarm.listen_on(/ip4/0.0.0.0/**tcp**/8080)
```

Hence, I think we should design the `TorTransport` in a way that it only does something if the user calls:

```
swarm.listen_on(/onion3/vww6ybal4bd7szmgncyruucpgfkqahzddi37ktceo3ah7ngmcopnpyyd:1234")
```

To make things work in the above case, the `Transport` would have to:

- assume that the hidden service is already set up
- use the control port to read information about the hidden service and figure out what port it forwards to locally
- listen on the local tcp port to make sure we receive the traffic coming in at this onion address

This will cover the usecase where the user wants to set up Tor outside of the application.

To make use of ephemeral services, we could take some inspiration from "wildcard" addresses: `0.0.0.0` is not actually a valid IP address and you can't listen on it. It is merely a placeholder that just shares the same syntax.

To convey the intent that the `TorTransport` should set up an ephemeral service, we could react to a special onion address in the form of:

```
swarm.listen_on(/onion3/00000000000000000000000000000000000000000000000000000000:1234")
```

Given a multi-address like this, the `TorTransport` would:

- listen on a TCP port locally
- generate a new key and set up an ephemeral service that forwards to the local TCP port
- emit a `NewListenAddr` event with the generated onion address to inform the rest of the application about it

There are a few more corner cases in regards to choosing ports etc but I think overall, this would be a much better design:

- We clearly separate ourselves from the `TcpTransport`, i.e. a `TorTransport` should not react to a `/ip4/0.0.0.0/tcp/0` multiaddress being passed to `swarm.listen_on`.
- We leave it completely up to the user to manage the hidden service themselves or having the application set up a hidden service.
- We can support dialling via Tor without having to listen on Tor.
- We don't have to filter addresses for localhost only: In the case of an ephemeral service, we can explicitly listen on `127.0.0.1`.

Reply:
Thanks for writing this up so clearly. Very Interesting idea. 
Some thoughts and questions from my side: 

> [...]
> ```
> swarm.listen_on(/onion3/vww6ybal4bd7szmgncyruucpgfkqahzddi37ktceo3ah7ngmcopnpyyd:1234")
> ```
> 

>   -  assume that the hidden service is already set up
>   -  use the control port to read information about the hidden service and figure out what port it forwards to locally
>   -  listen on the local tcp port to make sure we receive the traffic coming in at this onion address

From my current understand of Tor this feels a bit _weird_ because the application is not listening on this onion address. It's your local Tor instance which does that for you. 

The other thing is that I don't see a UX improvement with telling the application to listen on `/onion3/vww6ybal4bd7szmgncyruucpgfkqahzddi37ktceo3ah7ngmcopnpyyd:1234` vs telling it to listen on `/ipv/127.0.0.1/tcp/1234`. 
The user has to configure Tor in the first place knowing that the there is a service listening already or that he will start a service which listens on this address. To be precise, the user will need to add the following to their torrc file: 

```
HiddenServiceDir /usr/local/etc/tor/hidden_service
HiddenServicePort 3000 127.0.0.1:3000
```

>   -  use the control port to read information about the hidden service and figure out what port it forwards to locally

I would try to avoid any deep interaction with the control port. the control port spec are very verbose and if things fail it's not clear why. 
To give you a feeling, checkout the control port spec Section `3.9. GETINFO` from here https://gitweb.torproject.org/torspec.git/tree/control-spec.txt
In particular, these two requests seem to be relevant: 

```
    "hs/client/desc/id/<ADDR>"
      Prints the content of the hidden service descriptor corresponding to
      the given <ADDR> which is an onion address without the ".onion" part.
      The client's cache is queried to find the descriptor. The format of
      the descriptor is described in section 1.3 of the rend-spec.txt
      document.

      If <ADDR> is unrecognized or if not found in the cache, a 551 error is
      returned.

      [New in Tor 0.2.7.1-alpha]
      [HS v3 support added 0.3.3.1-alpha]

    "hs/service/desc/id/<ADDR>"
      Prints the content of the hidden service descriptor corresponding to
      the given <ADDR> which is an onion address without the ".onion" part.
      The service's local descriptor cache is queried to find the descriptor.
      The format of the descriptor is described in section 1.3 of the
      rend-spec.txt document.

      If <ADDR> is unrecognized or if not found in the cache, a 551 error is
      returned.

      [New in Tor 0.2.7.2-alpha]
      [HS v3 support added 0.3.3.1-alpha]
```

I'm not even sure what the responds will look like 😅


> To make use of ephemeral services, we could take some inspiration from "wildcard" addresses: 0.0.0.0 is not actually a valid IP address and you can't listen on it. It is merely a placeholder that just shares the same syntax.
> 
> To convey the intent that the TorTransport should set up an ephemeral service, we could react to a special onion address in the form of:
> 
> ```
> swarm.listen_on(/onion3/00000000000000000000000000000000000000000000000000000000:1234")
> ```

That's an interesting idea. Unfortunately parsing `/onion3/00000000000000000000000000000000000000000000000000000000` to a multi address will fail because it is not a valid multi-address format. It expects the onion address to be in base32 format, e.g. `oarchy4tamydxcitaki6bc2v4leza6v35iezmu2chg2bap63sv6f2did`. 

I understand that you would like to just have a _wild_card_ address here but I wouldn't know how a _nice_ wild_card address could look like. 

Imho a better UX would result from asking the user to explicitly state if she wants to have an hidden service registered for her or not.